### ISSUE TITLE
S2583 FP: Pattern matching generates wrong bool constraint

### ISSUE THREAD

**Issue by H4pH** [SonarSource/sonar-dotnet#3110]
### Description

False positive with pattern matching and bool value

### Repro steps

With this code, rule throw "Change this condition so that it does not always evaluate to 'true'; some subsequent code is never executed.". Value is `object`, and valueBool can be `true` or `false`, so conditional expression also can be `true` or `false`
```csharp
if (value == null || !(value is bool valueBool))
	return;

if (valueBool)
{
	// some code
}
else
{
	// some code
}
```

### Expected behavior

Don't throw

### Related information

* SonarC# Version 8.1
* Visual Studio Version 2019 
* MSBuild version 16
* If running through the Scanner for MSBuild, its version 4.7.1.2311-netcoreapp2.0

**Comment by pavel-mikula-sonarsource** [comment]
Hi @H4pH ,
Thank you for reporting this issue. We've confirmed it as FP on the latest SonarC# 8.3 as well.
